_posts/2020-8-17-the-new-scalable-ets-ordered_set.md

```diff
@@ -0,0 +1,245 @@
+   counter may eventually reach the threshold that indicates that the
+   base node should be split. A base node split divides the items in a
+   base node between two new base nodes and replaces the original base
+   node with a routing node where the two new base nodes are
+   rooted. The following picture shows the CA tree after the base node
+   pointed to by the tree's root has been split:
+   
+   
+   ![alt text](/images/ca_tree/ca_tree_2.png "First Split Contention Adapting Search Tree")
+3. The process of base node splitting will continue as long as there
+   is enough contention in base node locks or until the max depth of the
+   routing layer is reached. The following picture shows how the CA
+   tree looks like after another split:
+   
+   
+   ![alt text](/images/ca_tree/ca_tree_3.png "Second Split Contention Adapting Search Tree")
+4. The synchronization granularity may differ in different parts of a
+   CA tree if, for example, a particular part of a CA tree is accessed
+   more frequently in parallel than the rest. The following picture
+   shows the CA tree after yet another split:
+   
+   
+   ![alt text](/images/ca_tree/ca_tree_4.png "Third Split Contention Adapting Search Tree")
+5. The following picture shows the CA tree after the forth split:
+   
+   
+   ![alt text](/images/ca_tree/ca_tree_5.png "Forth Split Contention Adapting Search Tree")
+6. The following picture shows the CA tree after the fifth split:
+   
+   
+   ![alt text](/images/ca_tree/ca_tree_6.png "Fifth Split Contention Adapting Search Tree")
+7. Two base nodes holding adjacent ranges of items can be joined. Such
+   a join will be triggered after an operation sees that a base
+   node counter's value is below a certain threshold. Remember that a
+   base node's counter is decreased if a thread does not experience
+   contention when acquiring the base node's lock.
+   <!--The conters The likelihood that
+   a join will be triggered in a certain base node gets higher when
+   the probablity of contention that does not detect contention in the
+   base node lock is high. The likelihood that two base nodes are
+   joined is also increased if operations that require both base nodes
+   happens often enough (to reduce the overhead of acquiring locks).-->
+   
+   ![alt text](/images/ca_tree/ca_tree_7.png "Join of two base nodes in a  Contention Adapting Search Tree")
+8. As you might have noticed from the illustrations above, splitting
+   and joining results in that old base nodes and
+   routing nodes gets spliced-out from the tree. The memory that these
+   nodes occupy needs to be reclaimed, but this can not happen directly
+   after they have got spliced-out as some threads might still be
+   reading them. The Erlang run-time system has a mechanism called
+   [delayed
+   dealloc](https://github.com/erlang/otp/blob/d6285b0a347b9489ce939511ee9a979acd868f71/erts/emulator/internal_doc/DelayedDealloc.md),
```

Comment:
It's the "thread progress" mechanism that is used here to schedule deallocation jobs at a safe time.
"Delayed dealloc" is the allocator internal mechanism that sends back deallocation jobs to the thread that did the allocation. The word "delayed" can be misleading as that's not the purpose but rather a side effect.

Reply:
Thanks for the clarification! I have changed "delayed dealloc" in the text to "thread progress" and the link so that it points to the internal documentation for thread progress in the commit below.